### DISCUSSION TITLE
Removed System.Runtime.InteropServices.WindowsRuntime in .NET 5

### DISCUSSION THREAD

**Discussion by timaiv** [cefsharp/CefSharp#3360] [General]
From my custom app DependencyChecker `MetadataLoadContext.LoadFromAssemblyName()`
 "Could not find assembly 'System.Runtime.InteropServices.WindowsRuntime" (only 1 error, referenced from CefSharp.Core.Runtime)
 
System.Runtime.InteropServices.WindowsRuntime **is removed in .NET 5.**
https://blogs.windows.com/windowsdeveloper/2020/10/13/c-winrt-version-0-8-and-the-net5-rc2-release/
_We no longer project the System.Runtime.InteropServices.WindowsRuntime namespace, as it has been removed from .NET5._

I only found `using System.Runtime.InteropServices.WindowsRuntime`; in WPF example, so some other indirect reference.
__________
Also when i ported to .NET 5 my app i changed similar code (touch support) in my app to that:

        private static IInputPaneInterop GetIInputPaneInterop()
        {
            var guid = new Guid(IID_IInputPaneInterop);
            var name = StringToHString(RuntimeClass_Windows_UI_ViewManagement_InputPane);
            if (xWinInterop.Combase.RoGetActivationFactory(name, ref guid, out var factory) == 0)
            {
                return Marshal.GetObjectForIUnknown(factory) as IInputPaneInterop;
            }
            else
            {
                return null;
            }
        }

        public static InputPane GetInputPane(Window window)
        {
            try
            {
                // Need the HWND for the native interop call into IInputPaneInterop
                IntPtr handle = new WindowInteropHelper(window).Handle;

                var inputPaneComPtr = inputPaneInterop.GetForWindow(handle, typeof(IInputPane2).GUID);
                var inputPane = InputPane.FromAbi(inputPaneComPtr);
                return inputPane;
            }
            catch (Exception ex)
            {
                Debug.Fail(ex.Message);
                return null;
            }
        }

**Comment by amaitland** [comment]
> System.Runtime.InteropServices.WindowsRuntime is removed in .NET 5.
https://blogs.windows.com/windowsdeveloper/2020/10/13/c-winrt-version-0-8-and-the-net5-rc2-release/
We no longer project the System.Runtime.InteropServices.WindowsRuntime namespace, as it has been removed from .NET5.

I am aware the example needs updating.

> I only found using System.Runtime.InteropServices.WindowsRuntime; in WPF example, so some other indirect reference.

I'm not aware of any reference direct or indirect outside of the WPF example. 

You haven't specified which package or version you are testing with.

**Reply by amaitland** [reply]
A quick look at the `CefSharp.Core.Runtime.dll` provided in the `NetCore` packages and it would appear there is a reference to **every** library included in the `Microsoft.NETCore.App 3.1` framework (`%ProgramFiles%\dotnet\packs\Microsoft.NETCore.App.Ref\3.1.0\ref\netcoreapp3.1`). This would appear to be a bug in the `Net Core CLI/C++` implementation. Checking `CefSharp.dll` and it only has the references for the dlls that are used.

As we don't actually use `System.Runtime.InteropServices.WindowsRuntime` it's unlikely to cause a problem when running under `net5.0-windows`. I have yet to see an issue and have had no other reports. The dlls should be dynamically loaded, so under normal circumstances it will never be referenced.

Downloading the `Migration Example` from https://github.com/mjrousos/CppCliMigrationSample/tree/netcore (as referenced from https://devblogs.microsoft.com/cppblog/porting-a-c-cli-project-to-net-core/) and compiling the dll and it has the same resulting references. So this would appear to be a bug in `Net Core`.

I've asked at https://github.com/mjrousos/CppCliMigrationSample/issues/2 where to raise this problem.

**Reply by timaiv** [reply]
I will open some info. (not related to cef)

In .NET Core/.NET 5 more dll came to nuget, so file searching is more complicated (and has other problems, like `threadlocal` c++ on win 7 generates BadImageFormatException in Assemly.load (main problem, u can google it), or simpler: file renaming, creating empty file...). 
And System*.dll can be in runtime folder or in installed, we cant create some pattern matching. So i checked all dlls (in reflection mode only). 

Before my dependency checker also had other errors, some dlls are missed.
![image](https://user-images.githubusercontent.com/36565658/105465632-d5b49500-5c9b-11eb-965b-891dca931049.png)
Most of them moved to nuget, but they are still referenced in runtime folder))). So i set a check, if dll is inside core runtime i think all depencies already provided 

`
            if (Path.GetDirectoryName(assembly.Location) == Path.GetDirectoryName(RuntimeEnvironment.GetRuntimeDirectory()))
            {
                return;
            }
`

So after that only System.Runtime.InteropServices.WindowsRuntime is missed.

**Reply by amaitland** [reply]
We're not using the System.Runtime.InteropServices.WindowsRuntime, it's included by the compiler. Along with every framework dll.

![clicppnetcorerferences](https://user-images.githubusercontent.com/307872/105441595-a2323480-5cb4-11eb-8e05-a4822903fa4f.PNG)

If you can find a workaround then by all means let me know.

**Reply by amaitland** [reply]
I've opened https://github.com/dotnet/sdk/issues/15535 you are welcome to participate in the discussion there.
